test/performance/tools/common.sh

```diff
@@ -0,0 +1,143 @@
+function copy_secret() {
+  gcloud container clusters get-credentials ${MASTER_CLUSTER_NAME} --project=${PROJECT_NAME} --zone=${MASTER_CLUSTER_ZONE}
+  kubectl get secret service-account --context ${SOURCE_CONTEXT} --export -o yaml \
+    | kubectl apply --context $1 -f -
+}
+
+# Set up the user credentials for cluster operations.
+function setup_user() {
+  header "Setup User"
+
+  gcloud config set core/account ${USER_NAME}
+  gcloud auth activate-service-account ${USER_NAME} --key-file=${GOOGLE_APPLICATION_CREDENTIALS}
+  gcloud config set core/project ${PROJECT_NAME}
+
+  echo "gcloud user is $(gcloud config get-value core/account)"
+  echo "Using secret defined in $GOOGLE_APPLICATION_CREDENTIALS"
+}
+
+# Create a new cluster and install serving components and apply benchmark yamls
+# $1 -> cluster_name, $2 -> cluster_zone, $3 -> node_count
+function create_new_cluster() {
+  # create a new cluster
+  create_cluster $1 $2 $3
+  
+  # copy the secret to the new cluster
+  copy_secret "${CONTEXT_PREFIX}_${2}_${1}"
+
+  # update components on the cluster, e.g. serving and istio
+  update_cluster $1 $2
+}
+
+# Get serving code and apply the patch to it.
+function get_and_patch_serving() {
+  header "Apply hpa patch"
+  pushd .
+  mkdir -p ${GOPATH}/src/knative.dev
+  cd ${GOPATH}/src/knative.dev
+  git clone https://github.com/knative/serving.git
+
+  cd serving
+  echo "Using ko version $(ko version)"
+  echo "Patching autoscaler HPA replica count to 10"
+  git apply ${PROJ_ROOT_PATH}/tools/hpa_replicaCount.patch || fail_test "Failed to apply patch"
+  cd ..
+  popd
+}
+
+# Update resources installed on the cluster with the up-to-date code.
+# $1: name, $2: zone
+function update_cluster() {
+  name=$1
+  zone=$2
+  echo "Updating cluster with name ${name} in zone ${zone}"
+  gcloud container clusters get-credentials ${name} --zone=${zone} --project knative-performance || fail_test "Failed to get cluster creds"
+  
+  echo ">> Delete all existing jobs and test resources"
+  ko delete -f "${PROJ_ROOT_PATH}/${TEST_DIR}/$1"
+  kubectl delete job --all
+
+  pushd .
+  cd ${GOPATH}/src/knative.dev
+  echo ">> Update istio"
+  kubectl apply -f serving/third_party/istio-1.2-latest/istio-crds.yaml || fail_test "Failed to apply istio-crds"
+  kubectl apply -f serving/third_party/istio-1.2-latest/istio-lean.yaml || fail_test "Failed to apply istio-lean"
+
+  # Overprovision the Istio gateways.
+  kubectl patch hpa -n istio-system istio-ingressgateway \
+        --patch '{"spec": {"minReplicas": 10, "maxReplicas": 10}}'
+  kubectl patch deploy -n istio-system cluster-local-gateway \
+        --patch '{"spec": {"replicas": 10}}'
+
+  echo ">> Updating serving"
+  # Retry installation for at most two times as there can sometime be a race condition when applying serving CRDs
+  local n=0
+  until [ $n -ge 2 ]
+  do
+    ko apply -f serving/config/ -f serving/config/v1beta1 && break
```

Comment:
FWIW, if we had "hourly" builds, which published this and `ko resolve -f ./test/performance/config`, then this whole function could be applies and patches, which is trivial to port to golang.